models/montecarlo.py

```diff
@@ -0,0 +1,137 @@
+        self.cleaned_data = cleaned_data
+        self.epsilon = epsilon
+        self.state_action_dict = self.initialize_policy()
+
+    def initialize_policy(self):
+        return {(state, action): (0,0) 
+                for state in self.state_space
+                for action in range(len(self.action_space))}
+
+    def choose_action(self, state):
+        if np.random.rand() < self.epsilon:
+            return random.randint(0, len(self.action_space) - 1)  # Explore
+        return np.argmax(self.calculate_average_reward(state)) # Exploit
+
+    def update_policy(self, state, action, reward):
+        reward_sum, count = self.state_action_dict[(state, action)]
+        self.state_action_dict[(state, action)] = (round(reward_sum + reward, 2), count + 1)
+
+    def train_with_logs(self, episodes, train_env):
+
+        actions_taken  = []
+
+        for i in range(episodes):
+            episode_indexes  = self.generate_episode(train_env)
+
+            print("Starting episode : ", i)
+            print("____________________________________________________")
+
+            for t in episode_indexes:
+                state = train_env.get_discrete_state(index = t)
+                action_index = self.choose_action(state)
+                action = self.action_space[action_index]
+
+                print("Chosen State: ", state)
+                print("Chosen action: ", action)
+                actions_taken.append(action)
+
+                reward = train_env.get_reward(action, index =  t + 1)
+                print("reward this iter: ", reward)
+                self.update_policy(state, action_index, reward)
+
+                print("State action dict after iter: ")
+                print(self.pretty_print_state_action_dict(self.state_action_dict))
+                print("____________________________________________________")
+
+        print("State action dict after all episodes: ")
+        # print("actions taken : ", actions_taken)
+        print(self.pretty_print_state_action_dict(self.state_action_dict))
+
+        optimum_action_dict = {
+            state: np.argmax(self.calculate_average_reward(state))
+            for state in self.state_space
+        }
+
+        return optimum_action_dict
+    
+
+    def train(self, episodes, train_env):
+
+        for i in range(episodes):
+            episode_indexes  = self.generate_episode(train_env)
+
+            for t in episode_indexes:
+                state = train_env.get_discrete_state(index = t)
+                action_index = self.choose_action(state)
+                action = self.action_space[action_index]
+
+                reward = train_env.get_reward(action, index =  t + 1)
+                self.update_policy(state, action_index, reward)
+        
+        print("State action dict after all episodes: ")
+        print(self.pretty_print_state_action_dict(self.state_action_dict))
+
+        optimum_action_dict = {
```

Comment:
This can be a member data `self.optimum_action_dict` so that you don't need to pass it to the method `test`